Lib/tempfile.py

```diff
@@ -889,8 +890,31 @@ def resetperms(path):
 
                     try:
                         _os.unlink(path)
-                    # PermissionError is raised on FreeBSD for directories
-                    except (IsADirectoryError, PermissionError):
+                    except IsADirectoryError:
+                        cls._rmtree(path, ignore_errors=ignore_errors)
+                    except PermissionError:
+                        # The PermissionError handler was originally added for
+                        # FreeBSD in directories, but it seems that it is raised
+                        # on Windows too.
+                        # bpo-43153: Calling _rmtree again may
+                        # raise NotADirectoryError and mask the PermissionError.
+                        # So we must re-raise the current PermissionError if
+                        # path is not a directory.
+                        try:
+                            st = _os.lstat(path)
+                        except OSError:
+                            if ignore_errors:
+                                return
+                            raise
+                        if (_stat.S_ISLNK(st.st_mode) or
+                            not _stat.S_ISDIR(st.st_mode) or
+                            (hasattr(st, 'st_file_attributes') and
+                             st.st_file_attributes & _stat.FILE_ATTRIBUTE_REPARSE_POINT and
+                             st.st_reparse_tag == _stat.IO_REPARSE_TAG_MOUNT_POINT)
+                        ):
+                            if ignore_errors:
+                                return
+                            raise
                         cls._rmtree(path, ignore_errors=ignore_errors)
```

Comment:
Checking `S_ISLINK(mode)` is redundant here since the code also checks `not S_ISDIR(mode)`. The latter is always true for a symlink, as required by POSIX and as supported by Python on Windows. On the other hand, a junction (mount point) is always a directory on Windows.

Python 3.12+ has an optimized `os.path.isjunction()` function. It's always false on POSIX since POSIX doesn't have a junction file type. With the use of `isjunction()`, this check could be more simply written as follows:

```python
if not _os.path.isdir(path) or _os.path.isjunction(path):
    if ignore_errors:
        return
    raise
cls._rmtree(path, ignore_errors=ignore_errors)
```

Note that `os.unlink()` is implemented on Windows to delete a symlink or junction regardless of whether the target is a file or a directory. It should only fail with a permission error if the reparse point has the readonly file attribute set or if the effective security of the reparse point doesn't grant delete access to the current user.

Reply:
Thank you for your suggestion. Done in #112791.